lib/MySQL_Monitor.cpp

```diff
@@ -5816,5 +5846,1317 @@ bool DNS_Cache::empty() const {
+
+	short wait_events = 0;
+
+	if (status & MYSQL_WAIT_READ)
+		wait_events |= POLLIN;
+	if (status & MYSQL_WAIT_WRITE)
+		wait_events |= POLLOUT;
+	if (status & MYSQL_WAIT_EXCEPT)
+		wait_events |= POLLPRI;
+
+#ifdef DEBUG
+	int fd;
+
+	if (wait_events) {
+		fd = mysql_get_socket(mysql);
+	}
+	else {
+		fd = -1;
+	}
+	proxy_debug(PROXY_DEBUG_NET, 8, "fd=%d, wait_events=%d , old_ST=%d, new_ST=%d\n", fd, wait_events, async_state_machine_, new_st);
+#endif /* DEBUG */
+	
+	async_state_machine_ = new_st;
+	return wait_events;
+}
+
+static int
+mysql_status(short event) {
+	int status = 0;
+	if (event & POLLIN) 
+		status |= MYSQL_WAIT_READ;
+	if (event & POLLOUT) 
+		status |= MYSQL_WAIT_WRITE;
+	if (event & POLLPRI) 
+		status |= MYSQL_WAIT_EXCEPT;
+	return status;
+}
+
+class Monitor_Poll {
+public:
+	Monitor_Poll(unsigned int capacity) {
+		len_ = 0;
+		capacity_ = capacity;
+		fds_ = (struct pollfd*)malloc(capacity_ * sizeof(struct pollfd));
+		mmsds_ = (MySQL_Monitor_State_Data**)malloc(capacity_ * sizeof(MySQL_Monitor_State_Data*));
+	}
+
+	~Monitor_Poll() {
+		free(fds_);
+		free(mmsds_);
+	}
+
+	void expand(unsigned int more) {
+		if ((len_ + more) > capacity_) {
+			unsigned int new_size = near_pow_2(len_ + more);
+			fds_ = (struct pollfd*)realloc(fds_, new_size * sizeof(struct pollfd));
+			mmsds_ = (MySQL_Monitor_State_Data**)realloc(mmsds_, new_size * sizeof(MySQL_Monitor_State_Data*));
+			capacity_ = new_size;
+		}
+	}
+
+	void add(short _events, MySQL_Monitor_State_Data* mmsd) {
+		assert(mmsd->mysql);
+
+		if (len_ == capacity_) {
+			expand(1);
+		}
+		fds_[len_].fd = mysql_get_socket(mmsd->mysql);
+		fds_[len_].events = _events;
+		fds_[len_].revents = 0;
+		mmsds_[len_] = mmsd;
+		len_++;
+
+		mmsd->init_async();
+		mmsd->task_handler(-1, _events);
+	}
+
+	void remove_index_fast(unsigned int i) {
+		if ((int)i == -1) return;
+
+		if (i != (len_ - 1)) {
+			fds_[i].fd = fds_[len_ - 1].fd;
+			fds_[i].events = fds_[len_ - 1].events;
+			fds_[i].revents = fds_[len_ - 1].revents;
+			mmsds_[i] = mmsds_[len_ - 1];
+		}
+		len_--;
+	}
+
+	int find_index(int fd) {
+		unsigned int i;
+		for (i = 0; i < len_; i++) {
+			if (fds_[i].fd == fd) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+
+	bool event_loop(int poll_timeout_ms) {
+		
+		if (count() == 0)
+			return false;
+
+		int rc = 0;
+
+		while (len_ || GloMyMon->shutdown == true) {
+
+			rc = poll(fds_, len_, poll_timeout_ms);
+			
+			if (rc == -1) {
+				if (errno == EINTR) {
+					continue;
+				}
+				else {
+					return false;
+				}
+			}
+
+			for (unsigned int i = 0; i < len_;) {
+
+				if (mmsds_[i]->task_handler(fds_[i].revents, fds_[i].events) != MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_PENDING) {
+					remove_index_fast(i);
+					continue; 
+				}
+				else {
+					assert(fds_[i].events != 0);
+				}
+
+				fds_[i].revents = 0;
+				i++;
+			}
+
+		}
+
+		return true;
+	}
+
+	inline 
+	unsigned int count() const {
```

Comment:
This is just used once in the class itself, checking the private member should be enough.

Reply:
Done. This method can be useful in future so leaving it.

Reply:
Okay, it's small, I don't see the point. But we can leave it.